### PR TITLE
Bring user deletion account

### PR DESCRIPTION
**Pull request by tcitworld**
Bring the possibility for an user to delete it's own account. Administrators can enable/disable this functionality.

See https://github.com/owncloud/core/issues/158
- [ ] Security issues
- [ ] Redirect to login with js after deleting account
- [x] Tests

### CODE REVIEW

File: settings/Controller/UsersController.php [hunk 1]
```diff
@@ -433,27 +433,29 @@ public function create($username, $password, array $groups=array(), $email='') {
...
 	 * @NoAdminRequired
+	 * @NoSubadminRequired
 	 *
 	 * @param string $id
 	 * @return DataResponse
 	 */
 	public function destroy($id) {
 		$userId = $this->userSession->getUser()->getUID();
 		$user = $this->userManager->get($id);
+		$userOwnDeletion = $this->config->getSystemValue('user_own_account_deletion', false);
```
Comment (MorrisJobke) [inline settings/Controller/UsersController.php]: Please no config.php switch. This should be stored in the database.

File: settings/js/personal.js [hunk 2]
```diff
@@ -193,6 +193,21 @@ function avatarResponseHandler (data) {
 	}
 }
 
+function deleteAccount () {
```
Comment (MorrisJobke) [inline settings/js/personal.js]: Global namespace 🙈 Yes I know, the other methods are also in the global namespace - but at some point we need to clean up this mess 😉

Comment (MorrisJobke) [inline settings/js/personal.js]: Global namespace 🙈 Yes I know, the other methods are also in the global namespace - but at some point we need to clean up this mess 😉

**Comment by rullzer** [comment]
Can't we just do this as an app?

**Comment by tcitworld** [comment]
I thought it was a quite small feature to create an app. Also it's configurable by the admin.

File: settings/Controller/UsersController.php (hunk 1 above)
Comment (tcitworld) [inline settings/Controller/UsersController.php]: Okay. Where should it be activated by the administrator. Additional settings ?

**Review by LukasReschke — CHANGES_REQUESTED** [review]
I'd also prefer this to be in it's own app. But that may just be personal preference.

In any case, there are some problems with this:
- Missing unit tests for the UsersController
- The "delete your account" should only appear in the settings when it is possible (e.g. the user backend of the currently logged-in user supports it)
- Some remarks that Morris had as well.

**Comment by tcitworld** [comment]
Most of it is done. I just don't know where the admin setting should show up. Also, how to test if backend supports deletion (I've just put Database for now) ?

File: settings/templates/personal.php [hunk 3]
```diff
@@ -245,6 +245,13 @@
 	</div>
 </div>
 
+<?php if ($_['accountDeletionEnabled']) { ?>
```
Comment (nickvergessen) [inline settings/templates/personal.php]: Should be at the bottom of the site in my opinion

**Comment by LukasReschke** [comment]
> Most of it is done. I just don't know where the admin setting should show up. Also, how to test if backend supports deletion (I've just put Database for now) ?

`\OCP\UserInterface::implementsActions` + add an action constant

File: settings/personal.php [hunk 4]
```diff
@@ -142,6 +142,8 @@
 	$enableCertImport = $backendService->isUserMountingAllowed();
 }
 
+// show account deletion button only if it's enabled and if the backend supports it
+$accountDeletionEnabled = $config->getAppValue('core', 'user_own_account_deletion', false) && $user->getBackendClassName() === 'Database';
```
Comment (LukasReschke) [inline settings/personal.php]: > Most of it is done. I just don't know where the admin setting should show up. Also, how to test if backend supports deletion (I've just put Database for now) ?

`\OCP\UserInterface::implementsActions` + add an action constant + adjust the backends.

No hard-coded hackery here 😉

**Review by LukasReschke — CHANGES_REQUESTED** [review]

File: lib/private/User/User.php [hunk 5]
```diff
@@ -312,6 +312,19 @@ public function canChangeDisplayName() {
 	}
 
 	/**
+	 * check if the backend supports deleting user
+	 *
+	 * @return bool
+	 */
+	public function canDeleteAccount() {
+		// TODO : Change this
+		if (!filter_var($this->config->getAppValue('core', 'user_own_account_deletion', false), FILTER_VALIDATE_BOOLEAN)) {
```
Comment (tcitworld) [inline lib/private/User/User.php]: For some reason. Any help here ?

File: lib/public/IUser.php [hunk 6]
```diff
@@ -133,6 +133,14 @@ public function canChangePassword();
 	public function canChangeDisplayName();
 
 	/**
+	 * check if the backend supports deleting user
+	 *
+	 * @return bool
+	 * @since 11.0.0
```
Comment (tcitworld) [inline lib/public/IUser.php]: This or ownCloud notation ?

File: settings/personal.php [hunk 7]
```diff
@@ -214,5 +214,9 @@
 
 $formsAndMore = array_merge($formsAndMore, $formsMap);
 
+if ($user->canDeleteAccount()) {
```
Comment (tcitworld) [inline settings/personal.php]: Better way ?

File: lib/private/User/User.php (hunk 5 above)
Comment (LukasReschke) [inline lib/private/User/User.php]: `(bool)$this->config->getAppValue('core', 'user_own_account_deletion', 0)` and store an integer would have been my approach.

File: lib/public/IUser.php (hunk 6 above)
Comment (LukasReschke) [inline lib/public/IUser.php]: I'm fine with this.

File: settings/personal.php (hunk 7 above)
Comment (LukasReschke) [inline settings/personal.php]: 👍

**Comment by tcitworld** [comment]
Rebased. Please review. :-)

**Comment by rullzer** [comment]
I still really dislike this being in core. In 99.99% of the cases this is not required. For me the only real usecase is if a service provider wants to allow people to delete their accounts. Which is a fair use case. But, again, please do it as an app.

**Comment by MariusBluem** [comment]
> For me the only real usecase is if a service provider wants to allow people to delete their accounts. Which is a fair use case. But, again, please do it as an app.

Agreed. And Service Providers may want to use their own CRM to have better control over it. So the use-case is as small as this is something for a separate app, I think 😁 ...However: Good work 👍

**Comment by tcitworld** [comment]
Okay ! (I will delete the branch later)

File: settings/Controller/UsersController.php [hunk 8]
```diff
@@ -433,27 +433,29 @@ public function create($username, $password, array $groups=array(), $email='') {
...
 	 * @NoAdminRequired
+	 * @NoSubadminRequired
 	 *
 	 * @param string $id
 	 * @return DataResponse
 	 */
 	public function destroy($id) {
 		$userId = $this->userSession->getUser()->getUID();
 		$user = $this->userManager->get($id);
+		$userOwnDeletion = $this->config->getAppValue('core', 'user_own_account_deletion', false);
```
Comment (nickvergessen) [inline settings/Controller/UsersController.php]: default value should be `0`

File: settings/js/personal.js [hunk 9]
```diff
@@ -41,6 +41,21 @@ jQuery.fn.keyUpDelayedOrEnter = function (callback, allowEmptyValue) {
 	});
 };
 
+OC.Settings.Personal = OC.Settings.Personal || {
+	deleteAccount: function() {
+		OC.dialogs.confirm('Do you really want to delete your account ? All your data will be lost !', 'Delete your account ?', function (res) {
```
Comment (nickvergessen) [inline settings/js/personal.js]: Please remove the spaces before `?` and `!`

**Review by nickvergessen — COMMENTED** [review]
Needs changes

**Pull request closed without merging**